[14.0][ADD] DO NOT MERGE: pms_website_sale

carmenbianca opened the pull request:
Creating this here before creating a (potential) PR against upstream pms.

Commit:
🐛 fix(res_partner.py): use api.model decorator for create method to avoid pylint warning
🐛 fix(res_partner.py): use values instead of vals in create method to improve semantics

robinkeunen committed:
[REF] fix precommit

robinkeunen committed:
[REF] precommit: silence warning redundant-modulename-xml

robinkeunen committed:
[REF] precommit: silence warning on incoherent-interpreter-exec-perm

robinkeunen committed:
[REF] precommit : eslint

carmenbianca committed:
[UPD] OCA template

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

voltrans-it-johan committed:
[add] update service line day_qty by service qty

carmenbianca committed:
[ADD] pms_website_sale

[IMP] pms_website_sale: Show a very basic list of room types

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

[IMP] pms_website_sale: Filter improvements

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

[IMP] pms_website_sale: Images

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

[REF] pms_website_sale: Use booking engine in controller

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

[IMP] pms_website_sale: Page for room types

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

[IMP] pms_website_sale: Load /room page when not logged in

During this work, I also realised that pms.room.type is a product, so I
incorporated that design, which required some subsequent refactoring.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

[IMP] pms_website_sale: Add important comment

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

robinkeunen committed:
[FIX] pms: fix NumRoomsSelectionModel.booking_engine_id

robinkeunen committed:
[ADD] pms_w_s: rooms controller, templates and tests

robinkeunen commented:
I don't understand why my R4 commits w/ same hash than those in branch 14.0 are displayed here.

robinkeunen commented:
Hi @huguesdk , can you give us a first look at the work ? I'd love

- comments on the code
- point of views on the fixme and todo's left in the code by Carmen and I.

What has been done : 

- get route to rooms list (/rooms)
- template for rooms
- get route to room page (room)
- unit tests and _integration_ tests on the route

What is still broken : 

- we do nothing with input data
- the buttons lead to nowhere
- the date picker assets are loaded but it does not work yet

The task is [here](https://gestion.coopiteasy.be/web#id=10987&action=475&active_id=499&model=project.task&view_type=form&menu_id=)

huguesdk commented on pms/wizards/pms_booking_engine.py:
```diff
@@ -203,7 +203,7 @@ def _compute_availability_results(self):
                         room_type_id=room_type_iterator.id,
                         pricelist_id=record.pricelist_id.id,
-                    )
+                    )  # isn't this akin to a global variable ?
```
to answer the comment: `pms.property.availability` is a (non-stored) computed field that depends on several context variables. i think that this is correct, although maybe not that commonly seen.

robinkeunen replied:
remove this comment.

robinkeunen commented:
@huguesdk Rémy is back, he will be taking a look :-)

huguesdk commented:
oh, i had just started looking at it. i will just post my only comment so far. ☺️

robinkeunen committed:
[ADD] pms_w_s: booking page and controller

robinkeunen committed:
[IMP] pms_w_s: /rooms templates

robinkeunen committed:
[IMP] pms_w_s: use BookingEngineParser in /rooms

remytms committed:
[FIX] pms_w_s: simple date picker and filters

robinkeunen committed:
[ADD] pms_w_s: /booking/address form

remytms committed:
[IMP] pms_w_s: rooms sorting

remytms committed:
[IMP] pms_w_s: rooms date search

Input are now required, to avoid user entering only one date an not the
other.

A button has been added to reset the date search.

remytms committed:
[FIX] pms_w_s: rooms price sorting

remytms committed:
[IMP] pms_w_s: base mechanism to submit room booking

remytms committed:
[FIX] pms_w_s: rooms tests

remytms committed:
[ADD] pms_w_s: persist parser and book rooms

This commit contains a refactor of the BookingEngineParser.

robinkeunen committed:
[REF] pms_w_s: reorganise unit tests

robinkeunen committed:
[ADD] pms_w_s: payment pages

remytms committed:
[FIX] pms_w_s: various parser fixes

- Use SESSION_KEY when saving data into session
- Discussion about the add_room_request() method and other method
- Catch error when casting values from the formular

remytms committed:
[ADD] pms_w_s: delete item in booking page

remytms committed:
[FIX] pms_w_s: parser: getting partner from id

remytms committed:
[ADD] pms_w_s: booking: show total amount and back button

remytms committed:
[ADD] pms_w_s: errors management

Errors can now be shown in the log file and properly to the user.
On the /rooms page and /booking page

robinkeunen commented on pms/models/pms_room_type.py:
```diff
@@ -209,6 +209,6 @@ def create(self, vals):
         self.ensure_one()
         capacities = self.room_ids.filtered(
-            lambda r: r.pms_property_id.id == pms_property_id.id
+            lambda r: r.pms_property_id.id == pms_property_id  # do not commit
```
Remove this. PR there : https://github.com/OCA/pms/pull/227

robinkeunen commented on pms_website_sale/controllers/booking_controller.py:
```diff
@@ -0,0 +1,163 @@
+                    be_parser.del_room_request(post.get("delete"))
+                except ParserError as e:
+                    logger.error(e)
+                    errors.append(e.usr_msg)
```
@remytms Since this is not the same form, I would process it in another route. WDYT?

robinkeunen replied:
Or even use a delete request on `/rooms/<type_id>` or similar

remytms replied:
Technically, it's two different form. But it operates on the same object: adding, updating or deleting rooms on booking object. That's why I've put all the logic in the same route.

Also, it's easier as we do not need to redirect user to a different route (the route they use to submit the form) after processing the form.

robinkeunen commented on pms_website_sale/controllers/booking_engine_parser.py:
```diff
@@ -0,0 +1,165 @@
+
+    def save(self):
+        """Save data into session"""
+        self._session[BookingEngineParser.SESSION_KEY] = self.data
```
why does it need to be done explicitely ?

remytms replied:
I'm not quite sure. It feels me strange to have to save it explicitly, but testing without saving explicitly fails. I've not dive into this issue more than putting that monkey fix.

robinkeunen commented on pms_website_sale/controllers/booking_engine_parser.py:
```diff
@@ -0,0 +1,165 @@
+        try:
+            room_type_id = int(room_type_id)
+        except (ValueError, TypeError):
+            raise ParserError("Cannot proceed due to incorrect value")
```
Log the ValueError or TypeError message for easier developper debugging.

robinkeunen commented on pms_website_sale/controllers/booking_engine_parser.py:
```diff
@@ -0,0 +1,165 @@
+            room_type_id = int(room_type_id)
+        except (ValueError, TypeError):
+            raise ParserError("Cannot proceed due to incorrect value")
+        del_index = None
+        for index, room_request in enumerate(self.data["rooms_requests"]):
+            if room_request.get("room_type_id") == room_type_id:
+                del_index = index
+        if del_index is not None:
+            self.data["rooms_requests"].pop(del_index)
```
why not rather ?

```suggestion
        rooms_requests = self.data["rooms_requests"]
        rooms_requests = [rr for rr in rooms_requests if rr[room_type_id] != room_type_id]
        self.data["rooms_requests"] = rooms_requests
```
or
```suggestion
        self.data["rooms_requests"] = list(filter(
            lambda rr: rr[room_type_id] != room_type_id, self.data["rooms_requests"]
        ))
```

remytms replied:
Indeed nice suggestion. :)

robinkeunen commented on pms_website_sale/templates/pms_room_type_templates.xml:
```diff
@@ -0,0 +1,254 @@
+                            aria-label="Shopping cart"
+                            title="Shopping cart"
+                        >
+                            <span class="fa fa-shopping-cart" />
```
should be invisible if no room is selected

remytms replied:
I add that on the design task.

robinkeunen commented on pms_website_sale/controllers/booking_controller.py:
```diff
@@ -0,0 +1,163 @@
+                    be_parser.set_daterange(
+                        post.get("start_date"), post.get("end_date"), overwrite=False
+                    )
+                    be_parser.add_room_request(
+                        post.get("room_type_id"),
+                        post.get("quantity"),
+                        post.get("start_date"),
+                        post.get("end_date"),
+                    )
```
Only allows to process one room type at a time while the form allows to select multiple room types.

remytms replied:
That's a design issue. As is, the `rooms` page does not allows you to select several rooms in one shot.

If such a design is needed (selecting several rooms at once) than we should change the way we create forms in the `rooms` page.

robinkeunen commented on pms_website_sale/controllers/booking_engine_parser.py:
```diff
@@ -0,0 +1,165 @@
+            room_availability.value_num_rooms_selected = room["quantity"]
+
+    def _get_room_request(self, room_type_id):
+        """Return (first) room request that match room_type_id"""
```
I see a lot of looping on the requests, should we rather use a dictionary with room_type_id as keys for rooms requests ?

remytms replied:
Looks interesting indeed.

robinkeunen replied:
will  you do it ? Or should I ?

robinkeunen commented on pms_website_sale/controllers/room_controller.py:
```diff
@@ -0,0 +1,134 @@
+        # work.
+        be_parser = BookingEngineParser(request.env, {})
+        try:
+            be_parser.set_daterange(post.get("start_date"), post.get("end_date"))
```
Can't this be done in the parser's constructor (__init__)?

remytms replied:
I'm not in favor of that. We have a `set_` method for all operation and that seams coherent for me. Also `set_` method can trigger ParserError where the __init__ does not.

robinkeunen commented on pms_website_sale/controllers/room_controller.py:
```diff
@@ -0,0 +1,134 @@
+            "end_date": post.get("end_date"),
+            "availability_results": self._sort_availability_results(
+                booking_engine.availability_results, post.get("order")
+            ),
```
I'd find it more readable to sort outside the dictionary.

robinkeunen commented on pms_website_sale/controllers/room_controller.py:
```diff
@@ -0,0 +1,134 @@
+            # items. If you want this feature, it shouldn't be too difficult to
+            # write.
+            ("is_published", "=", True),
+        ]
```
remove for the moment

remytms replied:
I removed the functions as there are no longer used.

robinkeunen commented on pms_website_sale/controllers/room_controller.py:
```diff
@@ -0,0 +1,134 @@
+        order = post.get("order") or "name ASC"
+        return "%s, id desc" % order
+
+    def _sort_availability_results(self, records, order):
+        """Return sorted list of result based on order"""
+        if order == "name asc":
+            return records.sorted(
+                key=lambda r: r.room_type_id.name,
+                reverse=False,
+            )
+        elif order == "name desc":
+            return records.sorted(
+                key=lambda r: r.room_type_id.name,
+                reverse=True,
+            )
+        elif order == "list_price asc":
+            return records.sorted(
+                key=lambda r: r.price_per_room,
+                reverse=False,
+            )
+        elif order == "list_price desc":
+            return records.sorted(
+                key=lambda r: r.price_per_room,
+                reverse=True,
+            )
+        else:
+            return records
```
Suggestion, might me too hairy but I like it, I cringe on explosive if combinatory.

```suggestion
    def _sort_availability_results(self, records, order):
        """Return sorted list of result based on order"""
        key, *direction = order.split(" ")
        key_functions = {
            "name": lambda r: r.room_type_id.name,
            "list_price": lambda r: r.price_per_room,
        }
        key_function = key_functions.get(key, key_functions["name"])

        if direction:
            reverse = direction.pop() == "desc"
        else:
            reverse = False

        return records.sorted(
            key=key_function,
            reverse=reverse,
        )
```

remytms replied:
Really nice. :)

robinkeunen committed:
[IMP] pmw_w_s: remove is_published

robinkeunen committed:
[IMP] pws_w_s: improve room sort function

remytms committed:
[IMP] pms_w_s: address validation

robinkeunen committed:
[IMP] pms_w_s: process online payment

remytms committed:
[ADD] pms_w_s: booking/extra_info

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 82.1% (1478 of 1799 strings)

Translation: pms-14.0/pms-14.0-pms
Translate-URL: https://translation.odoo-community.org/projects/pms-14-0/pms-14-0-pms/es/

remytms committed:
[FIX] pms_w_s: tests

Using URLLookup `model()` does not work for public user as public user
has no specific rights on any object.

Another way to solve it is to give read rights for pblic user. But that
seams dangerous to me as any request made on the website is performed by
public user.

remytms committed:
[IMP] pms_w_s: error management and code readability

remytms committed:
[FIX] pms_w_s: fix price for rooms

Using `price_total` from the `availability_result` does not work as by
default in the booking.engine wizard the number of selected rooms
(`num_room_selected`) is equal to zero. Therefor the `price_total` is
null.

To solve that we set the amount of room selected to 1 for each rooms
available.

As the booking.engine for the rooms page is not stored this is a
temporary value that will not be used for booking.

remytms committed:
[ADD] pms_w_s: warn if daterange is different

remytms committed:
[FIX] pms_w_s: log form error as debug msg

remytms committed:
[IMP] pms_w_s: remove unused functions

remytms committed:
[FIX] pms_w_s: prefix routes to avoid conflict

Using route `/rooms` conflict with other pms modules.

Now routes for booking online are prefixed `/ebooking`.

remytms committed:
[ADD] pms_w_s: form submission /booking

robinkeunen committed:
[IMP] pms_w_s: process failed payments

robinkeunen commented on pms_website_sale/controllers/booking_controller.py:
```diff
@@ -0,0 +1,273 @@
+        website=True,
+        methods=["GET", "POST"],
+    )
+    def booking_address(self, **post):
```
@remytms in order to test the controller, you should only do the rendering in the route method. check `_booking_payment` and `test_booking_payment`.

robinkeunen commented on pms_website_sale/controllers/booking_engine_parser.py:
```diff
@@ -0,0 +1,238 @@
+    def set_partner(self, name, email, phone, address, city, postal_code, country_id):
+        """Add partner values. Values should be clean."""
+        if not name:
+            raise ParserError("Name is required.")
```
@remytms  Should be dealt with in the frontend

robinkeunen committed:
[IMP] pms_w_s: create partner in booking parser

robinkeunen committed:
[FIX] pms_w_s: room route

remytms committed:
[ADD] pms_w_s: rooms list design

robinkeunen committed:
[IMP] pms_w_s: catch availability errors

robinkeunen committed:
[REF] pmw_w_s: routes on top of class

robinkeunen committed:
[REF] pms_ws: split routes from processor

remytms committed:
[IMP] pms_w_s: design booking process pages

remytms committed:
[CHG] pms_w_s: rename template file and errors in page /booking

remytms committed:
[FIX] pms_w_s: tests for deleting room in booking

remytms committed:
[FIX] pms_w_s: booking delete button alignement

remytms committed:
[FIX] pms_w_s: availability redirection error

remytms committed:
[FIX] pms_w_s: partner creation

When testing to create a folio from an empty parser, the creation of the
partner fails because no partner data are present in the parser.

This fixes the tests, but it lead to issue in the case when the user
bypass the address formular in the booking process.

robinkeunen committed:
[IMP] pmw_w_s: tokenize /booking/success

robinkeunen committed:
[REF] pms_w_s: prefix functions w/ http verb

remytms committed:
[FIX] pms_w_s: raise error at partner creation

remytms committed:
[FIX] pms_w_s: AvailabilityError

remytms committed:
[FIX] pms_w_s: fix booking flow and notes

remytms committed:
[FIX] pms_w_s: rooms list images and reword

remytms committed:
[FIX] pms_w_s: center text when no rooms available

robinkeunen committed:
[ADD] pmw_w_s: terms and conditions

remytms committed:
[ADD] pms_w_s: info about tourist tax

remytms committed:
[FIX] pms_w_s: order minimum 1 room

DarioLodeiros committed:
[FIX]pms_l10n_es_sii: fix pms depends manifest

OCA-git-bot committed:
[UPD] README.rst

DarioLodeiros committed:
[IMP]pms: index m2o fields: https://github.com/odoo/odoo/pull/39332

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pms-14.0/pms-14.0-pms
Translate-URL: https://translation.odoo-community.org/projects/pms-14-0/pms-14-0-pms/

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pms-14.0/pms-14.0-pms_l10n_es
Translate-URL: https://translation.odoo-community.org/projects/pms-14-0/pms-14-0-pms_l10n_es/

DarioLodeiros committed:
[FIX]pms: compute_service_line_ids bug self use by record

DarioLodeiros committed:
[FIX]pms: compute price reservation line without company in cache

DarioLodeiros committed:
[FIX]pms: fix search property parent_id partner model related

DarioLodeiros committed:
[IMP]pms: improvement compute pricelist_id in folio and reservation

DarioLodeiros committed:
[FIX]pms: searchPanel delete domain multi

DarioLodeiros committed:
[FIX]pms: searchPanel delete property from view room search

DarioLodeiros committed:
[FIX]pms: compute_pricelist_id is_new instance wrong computed

DarioLodeiros committed:
[IMP]pms: untaxed_amount_to_invoice dont take account downpayments

DarioLodeiros committed:
[FIX]pms: boardservices with multi lines with the same product
